pil_analyzer/src/evaluator.rs

```diff
@@ -1,104 +1,394 @@
+        }
+    }
+
+    pub fn evaluate<'a, T: FieldElement, C: Custom>(
+        expr: &'a Expression<T>,
+        locals: &[Rc<Value<'a, T, C>>],
+        symbols: &impl SymbolLookup<'a, T, C>,
+    ) -> Result<Value<'a, T, C>, EvalError> {
+        Ok(match expr {
+            Expression::Reference(reference) => evaluate_reference(reference, locals, symbols)?,
+            Expression::PublicReference(name) => symbols.lookup_public_reference(name)?,
+            Expression::Number(n) => Value::Number(*n),
+            Expression::String(s) => Value::String(s.clone()),
+            Expression::Tuple(items) => Value::Tuple(
+                items
+                    .iter()
+                    .map(|e| evaluate(e, locals, symbols))
+                    .collect::<Result<_, _>>()?,
+            ),
+            Expression::ArrayLiteral(elements) => Value::Array(
+                elements
+                    .items
+                    .iter()
+                    .map(|e| evaluate(e, locals, symbols))
+                    .collect::<Result<_, _>>()?,
+            ),
+            Expression::BinaryOperation(left, op, right) => {
+                Value::Number(evaluate_binary_operation(
+                    evaluate_to_number(left, locals, symbols)?,
+                    *op,
+                    evaluate_to_number(right, locals, symbols)?,
+                ))
             }
-            Expression::PublicReference(r) => Err(format!("Cannot evaluate public reference: {r}")),
-            Expression::Number(n) => Ok(*n),
-            Expression::String(_) => Err("Cannot evaluate string literal.".to_string()),
-            Expression::Tuple(_) => Err("Cannot evaluate tuple.".to_string()),
-            Expression::ArrayLiteral(_) => Err("Cannot evaluate array literal.".to_string()),
-            Expression::BinaryOperation(left, op, right) => Ok(evaluate_binary_operation(
-                self.evaluate(left)?,
+            Expression::UnaryOperation(op, expr) => Value::Number(evaluate_unary_operation(
                 *op,
-                self.evaluate(right)?,
+                evaluate_to_number(expr, locals, symbols)?,
             )),
-            Expression::UnaryOperation(op, expr) => {
-                Ok(evaluate_unary_operation(*op, self.evaluate(expr)?))
+            Expression::LambdaExpression(lambda) => {
+                // TODO only copy the part of the environment that is actually referenced?
+                (Closure {
+                    parameter_count: lambda.params.len(),
+                    body: lambda.body.as_ref(),
+                    environment: locals.to_vec(),
+                })
+                .into()
             }
-            Expression::LambdaExpression(_) => {
-                Err("Cannot evaluate lambda expression.".to_string())
+            Expression::IndexAccess(index_access) => {
+                match evaluate(&index_access.array, locals, symbols)? {
```

Comment:
Is it a performance issue that we evaluate the entire array each time we access a single element? I'm not that deep into how arrays are actually used in practice, so maybe it's fine.

Reply:
I wondered about this as well, but I think this is also what you would expect from a programming language (except if you are used to lazy evaluation as default). If you call a function that returns an array, you expect this array to be "created" and are aware of the performance implications. Instead, if you already know that you just need a single element, you would not use an array but a function that implements index access.